File: backend/app/schemas/dns_record.py
"""DNS-record request and response schemas."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

# User-creatable record types. SOA is auto-managed by the zone bootstrap and
# cannot be created or edited via the records API, but it is returned in reads.
CreatableRecordType = Literal["A", "AAAA", "CNAME", "TXT", "MX", "NS", "PTR", "SRV", "CAA"]
RecordType = Literal["A", "AAAA", "CNAME", "TXT", "MX", "NS", "PTR", "SRV", "CAA", "SOA"]
RoutingPolicy = Literal["SIMPLE"]


class DnsRecordCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    type: CreatableRecordType
    ttl: int = Field(default=300, ge=0, le=604_800)
    value: str = Field(min_length=1)
    routing_policy: RoutingPolicy = "SIMPLE"


class DnsRecordUpdate(BaseModel):
    """All fields optional; the service applies a partial patch."""

    ttl: int | None = Field(default=None, ge=0, le=604_800)
    value: str | None = Field(default=None, min_length=1)


class DnsRecordRead(BaseModel):
    model_config = {"from_attributes": True}

    id: str
    hosted_zone_id: str
    name: str
    type: RecordType
    ttl: int
    value: str
    routing_policy: RoutingPolicy
    created_at: datetime
    updated_at: datetime

File: frontend/src/types/dns-record.ts
export const CREATABLE_RECORD_TYPES = [
  "A",
  "AAAA",
  "CNAME",
  "TXT",
  "MX",
  "NS",
  "PTR",
  "SRV",
  "CAA",
] as const;

export type CreatableRecordType = (typeof CREATABLE_RECORD_TYPES)[number];
export type RecordType = CreatableRecordType | "SOA";
export type RoutingPolicy = "SIMPLE";

export interface DnsRecord {
  id: string;
  hosted_zone_id: string;
  name: string;
  type: RecordType;
  ttl: number;
  value: string;
  routing_policy: RoutingPolicy;
  created_at: string;
  updated_at: string;
}

export interface DnsRecordCreatePayload {
  name: string;
  type: CreatableRecordType;
  ttl: number;
  value: string;
  routing_policy?: RoutingPolicy;
}

export interface DnsRecordUpdatePayload {
  ttl?: number;
  value?: string;
}

export interface ListRecordsParams {
  page?: number;
  page_size?: number;
  search?: string;
  type?: CreatableRecordType;
  sort?: string;
}
